config/doom/config.org
#+TITLE: Doom Emacs Configuration
#+DESCRIPTION: Personal configuration for Doom Emacs.
#+STARTUP: overview
#+PROPERTY: header-args :results silent
#+SETUPFILE: https://fniessen.github.io/org-html-themes/org/theme-readtheorg.setup

* Overview

** What is this?

This is my personal [[https://en.wikipedia.org/wiki/Literate_programming][literate]] configuration for [[https://github.com/hlissner/doom-emacs][Doom Emacs]], mostly focused towards Org-mode and programming, written in Org-mode.

** Why is there an =;;;###autoload= on top of =defun= calls?

You can find more information on this on [[https://github.com/hlissner/doom-emacs/blob/develop/docs/getting_started.org#autoloadel-or-autoloadel][Doom's Getting Started page]].

#+begin_quote
*Note* that the source blocks which contain these functions are tangled into a different
file, =autoload.el=.
#+end_quote

** Symbol name prefixes

Symbol (functions, variables etc) names are prefixed with either =ch/= or =contrib/=. These
work sort of like namespaces that contain my /additions/ to Emacs. This kind of
differentiation makes it easy to discover custom additions.

Functions that I've written myself to either enhance a built-in feature or add extra
functionality, are prefixed with =ch/=.

Functions that I've basically copied from others are prefixed with =contrib/=.

* Requirements

To use this configuration, there are a couple of things that should be installed on your
system (other than the [[https://github.com/hlissner/doom-emacs#prerequisites][Doom prerequisites]]). ~doom doctor~ will catch most of them, if not
all, but I'll mention them anyway:

+ =Source Code Pro=, =Fira Code= and =Hack= fonts.
+ A LaTeX compiler for rendering math.
+ ~sqlite3~ for org-roam.
+ ~jsbeautify~ for formatting JS/CSS/HTML code.
+ ~python-black~ for formatting Python code.
+ ~python-isort~ for sorting Python imports.
+ Maybe I'm forgetting something.. Will try to update!

* Package management

Doom's packaging system dictates that package management should be done in the
=$DOOMDIR/packages.el= file, using the ~package!~ macro.

#+begin_quote
*Note* that every source block with a ~(package! ...)~, ~pin!~ or ~unpin!~ call is tangled
to the =packages.el= file!
#+end_quote

First of all, let's ensure that =byte-compile= is disabled for the =packages.el= file.

#+begin_src emacs-lisp :tangle "packages.el"
;; -*- no-byte-compile: t; -*-
#+end_src

Doom pins every package to a specific commit, but there are some packages I would like to
always be updated:

#+begin_src emacs-lisp :tangle "packages.el"
(unpin! doom-modeline)
(unpin! doom-themes)
(unpin! vterm)
#+end_src

On the other hand, there are some packages I have no use for, so let's just disable them.

#+begin_src emacs-lisp :tangle "packages.el"
(package! evil-escape :disable t)
(package! evil-quick-diff :disable t)
(package! evil-traces :disable t)
(package! cuda-mode :disable t)
(package! demangle-mode :disable t)
(package! disaster :disable t)
(package! overseer :disable t)
(package! buttercup :disable t)
(package! opencl-mode :disable t)
(package! glsl-mode :disable t)
(package! company-glsl :disable t)
(package! flyspell-lazy :disable t)
#+end_src

* General configuration

Before going any further, let's enable =lexical binding= for the resulting =config.el=.
See this [[https://nullprogram.com/blog/2016/12/22/][blog post]] for more information on why =lexical binding= is used.

#+begin_src emacs-lisp
;;; config.el -*- lexical-binding: t; -*-
#+end_src

Doom comes with very sane defaults, but there are some things I would like to work a
bit differently.

** Start on fullscreen

I usually maximize the frame right after opening Emacs. Let's have that happen by default.

#+begin_src emacs-lisp
(add-to-list 'default-frame-alist '(fullscreen . maximized))
#+end_src

** Theme

I switch between themes more frequently than I would like to. Some of my favorites are:
Protesilaos Stavrou's [[https://protesilaos.com/modus-themes/][modus-themes]], all the [[https://github.com/purcell/color-theme-sanityinc-tomorrow][Tomorrow themes]] from Steve Purcell, =doom-nord=,
=doom-one= and =doom-gruvbox= (also their light variants).

#+begin_src emacs-lisp :tangle "packages.el"
(package! modus-operandi-theme)
(package! modus-vivendi-theme)
(package! color-theme-sanityinc-tomorrow)
#+end_src

#+begin_src emacs-lisp
(setq doom-theme 'doom-nord)

;; Simple tweaks to modus-themes.
(cond
 ((equal doom-theme 'modus-operandi)
  (setq modus-operandi-theme-distinct-org-blocks t
        modus-operandi-theme-scale-headings nil
        modus-operandi-theme-completions 'moderate
        modus-operandi-theme-visible-fringes 'subtle))
 ((equal doom-theme 'modus-vivendi)
  (setq modus-vivendi-theme-distinct-org-blocks t
        modus-vivendi-theme-scale-headings nil
        modus-vivendi-theme-completions 'moderate
        modus-vivendi-theme-visible-fringes 'subtle)))
#+end_src

** Fonts

I'm a big fan of the =Source Code Pro=, =Fira Code= and =Hack= fonts.

#+begin_src emacs-lisp
(setq doom-font (font-spec :family "Hack" :size 15)
      doom-big-font (font-spec :family "Hack" :size 18)
      doom-variable-pitch-font (font-spec :family "Hack" :size 17))
(setq doom-unicode-font (font-spec :name "DejaVu Sans Mono"))
#+end_src

** Modeline

I usually use the "light" variant of the modeline which is included in Doom, instead of
~doom-modeline~. I would like it to be a bit smaller, and stop showing
information about buffer encoding.

#+begin_src emacs-lisp
(when (featurep! :ui modeline +light)
  (setq +modeline-height 15
        +modeline-encoding nil))
#+end_src

Let's also make the same changes to =doom-modeline=, for the times I do use it.
Moreover, there is an issue which causes the right side of the modeline to cut off.
According to Doom's documentation, to fix it we can either:
1. Tweak ~all-the-icons-scale-factor~,
2. Add some padding to the modeline definition,
3. Use another font for the modeline, with a lower =:height=.


I choose the second option.

#+begin_src emacs-lisp
(after! doom-modeline
  (setq doom-modeline-height 15
        doom-modeline-buffer-encoding nil)

  (doom-modeline-def-modeline 'main
    '(bar matches buffer-info remote-host buffer-position parrot selection-info)
    '(misc-info minor-modes checker input-method buffer-encoding major-mode process vcs "  ")))
#+end_src

I also like having the time displayed in the modeline, but not the load average.

#+begin_src emacs-lisp
(after! time
  (setq display-time-default-load-average nil)
  (display-time-mode 1))
#+end_src

** Dashboard

The Dooom ASCII art that is shown in the dashboard looks very nice, but I think I prefer
the Emacs logo.

#+begin_src emacs-lisp
(setq fancy-splash-image (concat doom-private-dir "splash.png"))
#+end_src

Also, I kinda know all the shortcuts, so there is no need for the shortcut menu.

#+begin_src emacs-lisp
(remove-hook '+doom-dashboard-functions #'doom-dashboard-widget-shortmenu)
#+end_src

** Line numbers

Don't show line numbers. I don't really get any useful information by knowing on which
line I'm currently on, and the impact on performance is quite noticeable.

#+begin_src emacs-lisp
(setq display-line-numbers-type nil)
#+end_src

** Frame title

Use the buffer's name as the frame title.

#+begin_src emacs-lisp
(setq frame-title-format '("%b"))
#+end_src

** Enable disabled functions

Doom by default disables the ~narrow-to-region~ function, but it is kinda useful.

#+begin_src emacs-lisp
(put 'narrow-to-region 'disabled nil)
#+end_src

** Evil-mode

+ When exiting /insert mode/, the cursor moves one position backwards. Setting
  ~evil-move-cursor-back~ to ~nil~ "fixes" that.

+ When I split a window, I usually want to move to that new split.

+ Use Emacs keybindings in ~dired~.

#+begin_src emacs-lisp
(after! evil
  (setq
   evil-move-cursor-back nil
   evil-split-window-below t
   evil-vsplit-window-right t)
  (add-to-list 'evil-emacs-state-modes 'dired-mode))
#+end_src

Having to press 3 buttons (=C-w h=, =C-w l= etc.) to move between windows is a bit much, so
let's create some shorter keybindings, using the Meta key (=Alt=).

#+begin_src emacs-lisp
(map! (:after evil
       "M-h" 'nil
       "M-j" 'nil
       "M-k" 'nil
       "M-l" 'nil
       "M-q" 'nil
       :enm "M-h" #'evil-window-left
       :enm "M-j" #'evil-window-down
       :enm "M-k" #'evil-window-up
       :enm "M-l" #'evil-window-right
       :enm "M-q" #'evil-quit))
#+end_src

Use ~vim-surround~ operator keys for ~evil-surround~'s change & delete functionality.

#+begin_src emacs-lisp
(map! (:after evil
       (:map evil-operator-state-map
        "r" 'evil-surround-change
        "d" 'evil-surround-delete)))
#+end_src

Finally, use =j/k= instead of =gj/gk= to move between visual lines.

#+begin_src emacs-lisp
(map! (:after evil
       (:map evil-normal-state-map
        "j" 'evil-next-visual-line
        "k" 'evil-previous-visual-line)))
#+end_src

** Generic advices

When jumping between org-mode headings and git hunks, I would like the buffer to recenter.
This can be easily done with an =advice=.

#+begin_src emacs-lisp
(defadvice! ch--recenter-after-jump (&rest _args)
  "Recenter after jumps"
  :after '(org-next-visible-heading
           org-previous-visible-heading
           git-gutter:next-hunk
           git-gutter:previous-hunk)
  (when (get-buffer-window)
    (recenter)))
#+end_src

* Completion and searching

** Word completion with company

#+begin_src emacs-lisp
(after! company
  (setq company-idle-delay 0.5
        company-global-modes '(not shell-mode)
        company-minimum-prefix-length 1
        company-tooltip-align-annotations t
        company-dabbrev-downcase nil
        company-tooltip-limit 10)
  (when (featurep! :completion company +childframe)
    (setq company-box-doc-delay 6
          company-box-enable-icon nil))
  ;; Make aborting less annoying.
  (add-hook 'evil-normal-state-entry-hook #'company-abort))
#+end_src

** Minibuffer completion with ivy

What I really like about Emacs, is the ability to search and narrow candidates. You don't
have to know where things are, just how to search for them.

There are many completion frameworks available, like [[https://github.com/emacs-helm/helm][helm]], [[https://github.com/raxod502/selectrum][selectrum]], [[https://github.com/abo-abo/swiper][ivy/counsel/swiper]],
the built-in ~icomplete~ etc. I currently use ~ivy~, since that's the default in Doom.

*** Ivy configuration

Here are some useful tips for working with ivy. While in a minibuffer prompt, e.g., ~M-x~:

+ You can create a /buffer/ with the selection candidates using ~C-c C-o~.

+ You can search for something and then press ~S-Space~ to "lock" this search term, and use
  it as a filter. For example, if you type =recentf= in ~M-x~ and then press ~S-Space~, the
  results will be narrowed down to only those containing the word =recentf=.

+ Pressing ~M-j~ will copy the word under the cursor to the minibuffer.

+ There are times -usually when searching for documentation- when you want to run the command
  but keep the minibuffer open. You can do that with ~C-M-m~.
  You can be even more efficient and use ~C-M-n~ or ~C-M-p~ to move to the next or previous
  command respectively and run it, all at once, without leaving the minibuffer.

+ ~C-p~ will focus the prompt instead of the selection. Especially useful when trying to
  create a file with a name similar to an existing file.


And here are all the above and some extras, presented in a more "compact" way:

| Key     | Function                | Description                                |
|---------+-------------------------+--------------------------------------------|
| C-c C-o | ivy-ocur                | Place the candidates in a buffer.          |
| C-M-m   | ivy-call                | Run command, keeping minibuffer open.      |
| M-i     | ivy-insert-current      | Insert matching selection in the prompt.   |
| M-j     | ivy-yank-word           | Copy word-at-point in the minibuffer.      |
| S-SPC   | ivy-restrict-to-matches | Filter candidates according to the prompt. |

Finally, here is my configuration. Some things to note:

+ =SPC= is set so it does *not* complete in the minibuffer.
+ ~ivy-read-action-function~ is set so ~ivy~ actions are displayed using a hydra.
  Otherwise, in cases where there are many actions, they overflow the popup.

#+begin_src emacs-lisp
(map! (:map minibuffer-local-completion-map
       "SPC" 'nil))

(after! ivy
  (setq ivy-count-format "(%d/%d) "
        ivy-display-style 'fancy
        ivy-use-selectable-prompt t
        ivy-fixed-height-minibuffer nil
        ivy-read-action-function #'ivy-hydra-read-action))
#+end_src

*** Counsel configuration

+ I would like a nicer separator when browsing the kill-ring (~counsel-yank-pop~).

+ [[https://github.com/BurntSushi/ripgrep][ripgrep]], or ~rg~ in short, is way faster than ~grep~, so let's use it for everything search-related.

#+begin_src emacs-lisp
(after! counsel
  (setq counsel-yank-pop-preselect-last t
        counsel-yank-pop-separator "\n—————————\n")
  (setq counsel-fzf-cmd "rg --files --hidden --no-ignore --smart-case --glob !.git | fzf -f \"%s\"")
  (setq counsel-grep-base-command
        "rg -M 120 --line-number --smart-case --with-filename --color never --hidden --no-heading %s")
  (setq counsel-git-cmd "rg --files"))
#+end_src

Finally, some keybindings for easier access to commonly used functions.
#+begin_src emacs-lisp
(map! (:after counsel
       (:leader
        :desc "Ripgrep"
        "s R" 'counsel-rg
        "s-p" 'counsel-fzf)))
#+end_src

*** Keybindings for swiper

Kinda odd that these don't exist by default.

#+begin_src emacs-lisp
(after! ivy
  (map! (:leader
         :desc "Search all buffers"
         "s a" 'swiper-all
         :desc "Search all buffers for thing at point"
         "s A" 'swiper-all-thing-at-point)))
#+end_src

** imenu

~imenu~ comes built-in with Emacs and offers completion-based buffer navigation. It builds
an "index" of positions and then displays the list using a completion interface (in our case,
~ivy~).

~imenu~ integrates very nicely with org-mode, making it easy to navigate around large
org-files, but there is an issue. When the content of an Org file is folded, jumping to a
heading using ~imenu~ will not unfold the heading.

Let's create a function and a hook to fix this issue.

#+begin_src emacs-lisp :tangle "autoload.el"
;;;###autoload
(defun ch/imenu-reveal-content ()
  "Reveal header content after `imenu' jump."
  (cond
   ((and (eq major-mode 'org-mode)
         (org-at-heading-p))
    (org-show-entry)
    (org-reveal t))))
#+end_src

#+begin_src emacs-lisp
(after! imenu
  (add-hook 'imenu-after-jump-hook 'ch/imenu-reveal-content))
#+end_src

*** imenu-list

Having a permanently visible imenu index instead of the completion can be useful at times.
[[https://github.com/bmag/imenu-list][imenu-list]] provides a nice sidebar with the contents of ~imenu~'s index.

#+begin_src emacs-lisp :tangle "packages.el"
(package! imenu-list)
#+end_src

#+begin_src emacs-lisp
(use-package! imenu-list
  :after imenu
  :config

  (defun contrib/imenu-list-dwim (&optional arg)
    "Convenience wrapper for `imenu-list'.
Move between the current buffer and a dedicated window with the
contents of `imenu'.

The dedicated window is created if it does not exist, while it is
updated once it is focused again through this command.

With \\[universal-argument] toggle the display of the window."
    (interactive "P")
    (if arg
        (imenu-list-smart-toggle)
      (with-current-buffer
          (if (eq major-mode 'imenu-list-major-mode)
              (pop-to-buffer (other-buffer (current-buffer) t))
            (imenu-list)))))

  (map! :after imenu
        (:leader
         :desc "imenu-list-dwim"
         "s I" 'contrib/imenu-list-dwim)))
#+end_src

** isearch

The built-in ~isearch~ command is honestly top notch. Below is a "cheat-sheet" for the
~isearch~ keybindings I find most useful.

| Key bind | Description                  |
|----------+------------------------------|
| C-s C-w  | Search char or word at point |
| M-s o    | Run 'occur' on regexp        |
| M-s h r  | Highlight regexp             |
| M-s h u  | Undo the highlighting        |
| C-s M-r  | Toggle regexp search         |
| M-%      | Query replace                |
| C-M-%    | Query replace regexp         |

However, some tweaks are needed.

#+begin_src emacs-lisp
(after! isearch
  (setq search-highlight t
        search-whitespace-regexp ".*?"
        isearch-lazy-count t
        lazy-count-prefix-format nil
        lazy-count-suffix-format " (%s/%s)"
        isearch-yank-on-move 'shift
        isearch-allow-scroll 'unlimited))

(map! "C-S-s" 'isearch-backward
      :map isearch-mode-map
      "C-S-s" 'isearch-repeat-backward)
#+end_src

** Search enhancements

Replacing the symbol at point is a pretty common task, but by default, it requires two
steps. Let's define a function to make this task faster.

Some functions that help perform some common tasks faster. Those tasks are:

+ replace the symbol at point,
+ search in my dotfiles directory,
+ Create an =occur= buffer with all the URLs found in the current buffer.

#+begin_src emacs-lisp :tangle "autoload.el"
;;;###autoload
(defun ch/isearch-query-replace-symbol-at-point ()
  "Run `query-replace-regexp' for symbol at point."
  (interactive)
  (isearch-forward-symbol-at-point)
  (isearch-query-replace-regexp))

;;;###autoload
(defun ch/find-in-dotfiles ()
  "Open a file from my dotfiles via a fuzzy filename search."
  (interactive)
  (doom-project-find-file (expand-file-name "~/Projects/dotfiles/")))

;;;###autoload
(defun ch/occur-urls ()
  "Produce a list with all the URLs found in the current buffer."
  (interactive)
  (let ((urls browse-url-button-regexp))
    (occur urls "\\&")))
#+end_src

Keybindings for easier access.

#+begin_src emacs-lisp
(map! (:after isearch
       :desc "QUery replace symbol at point"
       "M-s %" 'ch/isearch-query-replace-symbol-at-point)
      (:leader
       :desc "Find file in dotfiles"
       "f t" #'ch/find-in-dotfiles))
#+end_src

* Project navigation & management with projectile

+ Don't require a project in order to work.
+ Use external tools for faster indexing.
+ Don't treat ignored directories as projects.

#+begin_src emacs-lisp
(after! projectile

  (defun ch/projectile-ignore-project-p (project-root)
    (string-match-p "/\\.emacs\\.d/\\.local/straight/repos" project-root))

  (setq
   projectile-require-project-root 'prompt
   projectile-indexing-method 'alien
   projectile-ignored-projects '("~/" "/tmp" "~/.emacs.d/.local/straight/repos" "~/.emacs/.local/straight/build")
   projectile-ignored-project-function #'ch/projectile-ignore-project-p))
#+end_src

* File management with dired

** Base dired configuration

The defaults are pretty much fine, but there are few extra things I find quite nice:

+ Copy and delete recursively.
+ Don't permanently delete files. Instead, send them to trash.
+ Configure listing switches. (~man ls~ for more information).
+ Hide details like permissions, size etc by default.
+ Highlight the current line.

#+begin_src emacs-lisp
(after! dired
  (setq dired-recursive-deletes 'always
        dired-recursive-copies 'always
        delete-by-moving-to-trash t
        dired-listing-switches "-AGFhlv --group-directories-first --time-style=long-iso")
  (add-hook 'dired-mode-hook (lambda ()
                               (dired-hide-details-mode 1)
                               (hl-line-mode))))
#+end_src

** Extra functionality

*** dired-x

#+begin_src emacs-lisp
(after! dired-x
  (setq dired-clean-up-buffers-too t
        dired-clean-confirm-killing-deleted-buffers t
        dired-bind-info nil
        dired-bind-man nil))
#+end_src
*** dired-async

Gives ~dired~ the ability to perform actions asynchronously.

#+begin_src emacs-lisp
(after! (:and dired async)
  (dired-async-mode 1))
#+end_src
*** dired-open

~dired-open~ allows us to specify the default applications to be used for different
mime-types, when opening them from inside ~dired~.

#+begin_src emacs-lisp :tangle "packages.el"
(package! dired-open)
#+end_src

#+begin_src emacs-lisp
(use-package! dired-open
  :after (dired)
  :config
  (setq dired-open-extensions
        '(("pdf" . "evince")
          ("jpg" . "eog")
          ("png" . "eog")
          ("mkv" . "mpv")
          ("mov" . "mpv")
          ("mp3" . "mpv")
          ("mp4" . "mpv")
          ("avi" . "mpv"))))
#+end_src
*** dired-subtree

By default, ~dired~ provides a flat view of directories and has no concept of subtrees.
I really can't get used to that. ~dired-subtree~ brings subtrees to dired.

#+begin_src emacs-lisp :tangle "packages.el"
(package! dired-subtree)
#+end_src

#+begin_src emacs-lisp
(use-package! dired-subtree
  :after dired
  :config
  (setq dired-subtree-cycle-depth 4)
  (map! :map dired-mode-map
        [backtab] #'dired-subtree-cycle
        [tab] #'dired-subtree-toggle
        :n "g^" #'dired-subtree-beginning
        :n "g$" #'dired-subtree-end
        :n "gm" #'dired-subtree-mark-subtree
        :n "gu" #'dired-subtree-unmark-subtree))
#+end_src
*** dired-narrow

~dired-narrow~ allows us to search and narrow based on regex, with can be super useful
when browsing large directories.

#+begin_src emacs-lisp :tangle "packages.el"
(package! dired-narrow)
#+end_src

#+begin_src emacs-lisp
(use-package! dired-narrow
  :after dired
  :config
  (setq dired-narrow-exit-when-one-left t
        dired-narrow-enable-blinking t
        dired-narrow-blink-time 0.3)
  (map! :map dired-mode-map
        "C-c /" 'dired-narrow-regexp))
#+end_src

** Keybindings for dired

#+begin_src emacs-lisp
(map! :after dired
      (:map dired-mode-map
       :enm "M-h" #'evil-window-left
       :enm "M-j" #'evil-window-bottom
       :enm "M-k" #'evil-window-up
       :enm "M-l" #'evil-window-right
       :enm "M-q" #'evil-quit
       "j" 'dired-next-line
       "k" 'dired-previous-line
       "h" 'dired-up-directory
       "l" 'dired-open-file)
      "s-j" 'dired-jump
      "s-J" 'dired-jump-other-window
      "M-RET" 'dired-display-file)
#+end_src

* Buffer management

** Buffer cleanup

Here are some functions that can be used to:

+ Kill all the buffers except the current one.
+ Kill all dired buffers.

#+begin_src emacs-lisp :tangle "autoload.el"
;;;###autoload
(defun ch/kill-other-buffers ()
  "Kill all buffers except current one."
  (interactive)
  (mapc 'kill-buffer (delq (current-buffer) (buffer-list))))

;;;###autoload
(defun ch/kill-all-dired-buffers ()
  "Kill all open dired buffers."
  (interactive)
  (mapc (lambda (buffer)
          (when (eq 'dired-mode (buffer-local-value 'major-mode buffer))
            (kill-buffer buffer)))
        (buffer-list)))
#+end_src

Let's add some keybindings for those.
#+begin_src emacs-lisp
(map! "s-k" 'ch/kill-other-buffers
      "s-K" 'ch/kill-all-dired-buffers)
#+end_src

* Window navigation and management

** Fast window navigation with ace-window

#+begin_src emacs-lisp :tangle "packages.el"
(package! ace-window)
#+end_src

#+begin_src emacs-lisp
(use-package! ace-window
  :defer t
  :config
  (setq aw-keys '(?h ?j ?k ?l ?y ?u ?i ?o ?p))
  (setq aw-scope 'frame)
  (setq aw-dispatch-always t)
  (setq aw-dispatch-alist
        '((?s aw-swap-window "Swap Windows")
          (?2 aw-split-window-vert "Split Window Vertically")
          (?3 aw-split-window-horz "Split Window Horizontally")
          (?? aw-show-dispatch-help)))
  (setq aw-minibuffer-flag t)
  (setq aw-ignore-current nil)
  (setq aw-display-mode-overlay t)
  (setq aw-background t)
  (ace-window-display-mode -1))

(map! :leader
      (:prefix ("w" . "+window")
       :desc "Ace window" "a" 'ace-window))
#+end_src

** Popup rules

#+begin_src emacs-lisp
(set-popup-rule! "*eww*" :side 'bottom :size .70 :slect t :vslot 5 :ttl 3 :quit nil)
(set-popup-rule! "*Help.*" :side 'right :width .50 :select t :vslot 0 :ttl 3 :quit nil)
(set-popup-rule! "*Man.*" :side 'right :width .50 :select t :slot 1 :ttl 3 :quit 'nil :modeline nil)
(set-popup-rule! "*ansible-doc.*" :side 'bottom :height .40 :select t :vslot 1 :ttl 3 :quit nil)
(set-popup-rule! "*Flycheck errors*" :side 'bottom :height .2 :select nil :quit t)
(after! imenu-list
  (set-popup-rule! "^\\*Ilist"
    :side 'right :size 35 :quit nil :select t :ttl 0))
(set-popup-rule! "*helpful.*" :side 'bottom :height .30 :select t :vslot 0 :ttl 3 :quit 'other)
#+end_src

** Window related enhancements

There are times when I have multiple windows open at the same time, and I constantly want
to switch between the same /two/. ~other-window~ (~C-x o~) will cycle between every window, so
that's not an option.

The ~ch/switch-to-last-window~ function will switch between the two most recently used
windows, which is exactly what I want.

#+begin_src emacs-lisp :tangle "autoload.el"
;;;###autoload
(defun ch/switch-to-last-window ()
  "Switch to the most recently used window."
  (interactive)
  (let ((win (get-mru-window nil t t)))
    (unless win (error "Most recent window not found."))
    (select-window win)))
#+end_src

Let's also create a keybinding for it.

#+begin_src emacs-lisp
(map! "M-o" 'ch/switch-to-last-window)
#+end_src

* Editing configuration

** Buffer formatting

I like having buffers formatted on-save, but not for every mode.

#+begin_src emacs-lisp
(setq +format-on-save-enabled-modes
      '(not tex-mode
            latex-mode
            emacs-lisp-mode
            c-mode
            yaml-mode))
#+end_src

** Indentation

There is this [[https://dougie.io/emacs/indentation/#functions-for-enablingdisabling-tabs][great article]] explaining everything about indentation in Emacs.

#+begin_src emacs-lisp

config/doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;;; Appearance

;;;; Font configuration

(when window-system
  (if (<= (display-pixel-width) 1366)
      (setq doom-font (font-spec :family "Fira Code" :size 13)
            doom-big-font (font-spec :family "Fira Code" :size 15)
            doom-variable-pitch-font (font-spec :family "FiraGO" :size 14))
    (setq doom-font (font-spec :family "Fira Code" :size 15)
          doom-big-font (font-spec :family "Fira Code" :size 18)
          doom-variable-pitch-font (font-spec :family "FiraGO" :size 17))))
(setq doom-unicode-font (font-spec :name "DejaVu Sans Mono"))

;;;; Theme configuration
(setq doom-theme 'sanityinc-tomorrow-eighties)

;;;;; Modus-themes tweaks
(if (equal doom-theme 'modus-operandi)
    (setq modus-operandi-theme-distinct-org-blocks t
          modus-operandi-theme-scale-headings nil
          modus-operandi-theme-completions 'moderate
          modus-operandi-theme-visible-fringes 'subtle))

(if (equal doom-theme 'modus-vivendi)
    (setq modus-vivendi-theme-distinct-org-blocks t
          modus-vivendi-theme-scale-headings nil
          modus-vivendi-theme-completions 'moderate
          modus-vivendi-theme-visible-fringes 'subtle))

;;;; Don't show line numbers.
(setq display-line-numbers-type nil)

;;;; Show the default Emacs logo in the dashboard.
(setq fancy-splash-image (concat doom-private-dir "splash.png"))
;;;; Don't show the shortcut menu in the dashboard.
(remove-hook '+doom-dashboard-functions #'doom-dashboard-widget-shortmenu)

;;;; Use the buffer's name as the frame title.
(setq frame-title-format '("%b"))

;;;; Built-in modeline tweaks
;;;;; Smaller height for modeline +light
(when (featurep! :ui modeline +light)
  (setq +modeline-height 20
        +modeline-encoding nil))

;;;;; When using 'doom-modeline', don't show the buffer encoding.
(after! doom-modeline
  (setq doom-modeline-buffer-encoding nil))

;;;;; Use smaller all-the-icons icons.
;; This helps fix padding issues where doom-modeline "overflows" from
;; the right.
(setq all-the-icons-scale-factor 1.1)

;;;;; Always show the time in the modeline, but not the load average.
(after! time
  (setq display-time-default-load-average nil)
  (display-time-mode 1))

;;; Better defaults

;;;; Indentation settings
(setq-default tab-always-indent 'complete ;; Indent first, complete later
              indent-tabs-mode nil        ;; Indent with spaces
              tab-width 4)                ;; Use 4 spaces/TAB

;;;; Scrolling tweaks

;; Page scrolling should keep the point at the same visual position,
;; rather than force it to the top or bottom of the viewport.
(setq-default scroll-preserve-screen-position t)
(setq-default scroll-conservatively 1) ; affects `scroll-step'
(setq-default scroll-margin 0)

;;;; Enable disabled functions

;; By default, Doom disables the 'narrow-to-region' function.
;; I find it quite useful.
(put 'narrow-to-region 'disabled nil)

;;;; Associate filetypes with modes
(add-to-list 'auto-mode-alist '("\\.cheat\\'" . sh-mode))
(add-to-list 'auto-mode-alist '("\\zshenv" . sh-mode))

;;;; Emacsclient
;; Load "heavy" packages during daemon startup.
(defun greedy-daemon-startup ()
  (when (daemonp)
    (require 'org)
    (require 'org-roam)
    (require 'org-super-agenda)))

(add-hook 'emacs-startup-hook #'greedy-daemon-startup)

;;; Projectile configuration

;; 1. Don't require a project for projectile to work.
;; 2. Faster indexing.
;; 3. Don't create projects in the ignored directories.
(after! projectile

  (defun +projectile-ignore-project-p (project-root)
    (string-match-p "/\\.emacs\\.d/\\.local/straight/repos" project-root))

  (setq
   projectile-require-project-root 'prompt
   projectile-indexing-method 'alien
   projectile-ignored-projects '("~/" "/tmp" "~/.emacs.d/.local/straight/repos" "~/.emacs/.local/straight/build")
   projectile-ignored-project-function #'+projectile-ignore-project-p))

;;; Flycheck configuration

;; Don't enable flycheck in org-mode buffers.
;; Use 'symbols' as the highlighting method.
;; Note that 'lines' can be (way) faster, but 'symbols' behaves
;; like most modern editors.
(after! flycheck
  (setq flycheck-highlighting-mode 'symbols
        flycheck-global-modes '(not org-mode)))

;;; ibuffer configuration

(after! ibuffer
  (setq ibuffer-display-summary nil)
  (setq ibuffer-formats
        '((mark modified read-only locked " "
                (name 30 30 :left :elide)
                " "
                (mode 16 16 :left :elide)
                " " filename-and-process)
          (mark " "
                (name 16 -1)
                " " filename))))

;;; isearch configuration

(after! isearch
  (setq search-highlight t
        search-whitespace-regexp ".*?"
        isearch-lazy-count t
        lazy-count-prefix-format nil
        lazy-count-suffix-format " (%s/%s)"
        isearch-yank-on-move 'shift
        isearch-allow-scroll 'unlimited)

  (map! :desc "Query replace symbol at point"
        "M-s %" 'util/isearch-query-replace-symbol-at-point))

;;; yasnippet configuration

;; Use the same indentation as the snippet itself.
(after! yasnippet
  (setq yas-indent-line 'fixed
        yas-also-auto-indent-first-line t))

;;; highlight-indent-guides configuration

;; Disable indent-guides everywhere but yaml and python modes.
(after! highlight-indent-guides
  (setq highlight-indent-guides-method 'character
        highlight-indent-guides-auto-odd-face-perc 15
        highlight-indent-guides-auto-even-face-perc 15
        highlight-indent-guides-auto-character-face-perc 20))

(remove-hook 'text-mode-hook #'highlight-indent-guides-mode)
(remove-hook 'prog-mode-hook #'highlight-indent-guides-mode)
(add-hook 'yaml-mode-hook #'highlight-indent-guides-mode)
(add-hook 'python-mode-hook #'highlight-indent-guides-mode)

;;; format configuration
;; Disable format-on-save for specified modes.
(setq +format-on-save-enabled-modes
      '(not tex-mode
            latex-mode
            emacs-lisp-mode
            c-mode))

;;; Popup rules

;; TODO: There might be a way to autoload those rules, for better performance.
(set-popup-rule! "*eww*" :side 'bottom :size .70 :slect t :vslot 5 :ttl 3 :quit nil)
(set-popup-rule! "*Help.*" :side 'right :width .50 :select t :vslot 0 :ttl 3 :quit nil)
(set-popup-rule! "*Man.*" :side 'right :width .50 :select t :slot 1 :ttl 3 :quit 'current :modeline nil)
(set-popup-rule! "*ansible-doc.*" :side 'bottom :height .40 :select t :vslot 1 :ttl 3 :quit nil)
(set-popup-rule! "*Flycheck errors*" :side 'bottom :height .2 :select nil :quit t)
(after! imenu-list
  (set-popup-rule! "^\\*Ilist"
    :side 'right :size 35 :quit nil :select t :ttl 0))
(set-popup-rule! "*helpful.*" :side 'bottom :height .30 :select t :vslot 0 :ttl 3 :quit 'other)

;;; Run all Doom hooks on startup.

;; This will delay startup a bit, but it can offer a much
;; more responsive experience.
(run-hooks 'doom-first-input-hook)

;;; imenu

(after! imenu
  (add-hook 'imenu-after-jump-hook 'util/imenu-reveal-entry))

;;; Spelling

;; Use ispell's multiple dictionaries functionality.
(after! ispell
  (setq ispell-program-name "hunspell"
        ispell-dictionary "en_US,el_GR")
  (ispell-set-spellchecker-params)
  (ispell-hunspell-add-multi-dic "en_US,el_GR")
  (setq ispell-personal-dictionary "~/.hunspell_personal")

  (unless (file-exists-p ispell-personal-dictionary)
    (write-region "" nil ispell-personal-dictionary nil 0))

  ;; Don't spell check the following org source blocks.
  (add-to-list 'ispell-skip-region-alist '("#\\+BEGIN_SRC" . "#\\+END_SRC"))
  (add-to-list 'ispell-skip-region-alist '("#\\+BEGIN_LATEX" . "#\\+END_LATEX"))
  (add-to-list 'ispell-skip-region-alist '("#\\+BEGIN_EXAMPLE" . "#\\+END_EXAMPLE")))
